### PR TITLE
Add wurf android copy lib to app

### PR DESCRIPTION
**Pull request by jpihl**
@mortenvp @petya2164 any comments?

### CODE REVIEW

File: wurf_android_copy_lib_to_app.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    input_librarys = self.link_task.outputs
+    output_librarys = []
+    for input_library in input_librarys:
+        project_path = self.path.parent.abspath()
+        output_library = os.path.basename(input_library.abspath())
+        output_library = self.bld.root.make_node(
+            os.path.abspath(os.path.join(
+                project_path,
+                'app',
```
Comment (petya2164) [inline wurf_android_copy_lib_to_app.py]: So this path is not always accurate. We can put the "app" in any folder, so this relative path should a parameter in the wscript_build.

File: wurf_android_copy_lib_to_app.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    Copy library to Android application.
+
+    When building an Android application from the IDE native libraries placed
+    in the libs folder will be packaged in the apk. So we copy the shared libs
+    we build there.
+    """
+    if not self.bld.is_mkspec_platform('android'):
+        return
+
+    input_librarys = self.link_task.outputs
```
Comment (petya2164) [inline wurf_android_copy_lib_to_app.py]: libraries ;)
Actually, I like input_libs better! We should not overuse the no-abbreviation policy, where the meaning is absolutely clear.

Comment (jpihl) [inline wurf_android_copy_lib_to_app.py]: Doh! I also thought it looked a bit funny.. :)

File: wurf_android_copy_lib_to_app.py (hunk 1 above)
Comment (jpihl) [inline wurf_android_copy_lib_to_app.py]: how do you set such a parameter?

Comment (petya2164) [inline wurf_android_copy_lib_to_app.py]: I think we can set any parameter when we define the task:
```
bld(features='cxx cxxshlib android_copy_lib_to_app',
    source='native_media_receiver.cpp',
    target='native_media_receiver',
    export_includes=['jni'],
    install_path=None,
    copy_path='../app/src/main/jniLibs/armeabi',
    lib=['android'],
    use=['wurf_it'])
```
We also need to figure out if we want a relative path from the wscript_build, or from the project root.

Comment (jpihl) [inline wurf_android_copy_lib_to_app.py]: it seems `export_includes` has paths relative to the project root, so I would go with the same approach.

Comment (jpihl) [inline wurf_android_copy_lib_to_app.py]: Btw, the final folder `armeabi` is dependent on the mkspec. I'm not entirely sure how to map the cxx_mkspec to such a folder name, but I guess we can leave that for an other time.

Comment (petya2164) [inline wurf_android_copy_lib_to_app.py]: I found the relevant description for `export_includes`: https://github.com/waf-project/waf/blob/9625343f91810ffeb8b844f14f997daf44a91ff5/waflib/Tools/ccroot.py#L66
Clearly, the paths are relative and we have a mistake here: the wscript_build sits in a subfolder, so the path should be `../jni` or just `.` (but nobody uses the exported includes, so it does not matter here).

We should make the copy_path relative and we can use `self.path.make_node(os.path.join(self.copy_path, output_library))` to get the output node.

I should note that the resolution of project_path is also bogus here:
`project_path = self.path.parent.abspath()`
This only works if we are in a direct subfolder of the project root.
The correct solution would be: `self.bld.srcnode`, but we don't need this with the relative paths!

**Comment by petya2164** [comment]
@jpihl After creating the `copy_path` parameter, there is nothing Android-specific about this tool. We take the output of the link task (that can be any executable or shared lib), and copy it to some destination.
I think we should call the feature `copy_binary` and the helper class should be `CopyFileTask`. We can also remove all references to Android. Then we have a generic tool that can be used anywhere for a similar purpose ;)

File: wurf_android_copy_lib_to_app.py (hunk 1 above)
Comment (jpihl) [inline wurf_android_copy_lib_to_app.py]: So.. 
The export_includes should be removed.
copy_path should be relative.

File: wurf_android_copy_lib_to_app.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    Copy library to Android application.
+
+    When building an Android application from the IDE native libraries placed
+    in the libs folder will be packaged in the APK. So we copy the shared libs
+    we build there.
+    """
+    if not self.bld.is_mkspec_platform('android'):
+        return
+
+    if 'copy_path' not in dir(self):
```
Comment (jpihl) [inline wurf_android_copy_lib_to_app.py]: hasattr is better.

**Comment by jpihl** [comment]
I like making this a more generic approach. But that would make a a solution that automatically finds the final folder (armeabi, x86, etc.) do you have an idea for solving that?

**Comment by petya2164** [comment]
First of all, that is not relevant at this point.
And if we ever get to that problem, the solution would not affect this generic tool:
You could define some variable based on the current mkspec: `env['NDK_ARCH']="armeabi"`
then you can use that var in your copy_path:
`copy_path='../app/src/main/jniLibs/'+bld.env['NDK_ARCH']`

**Pull request merged**